ports/rp2/mphalport.c

```diff
@@ -199,17 +198,31 @@ mp_uint_t mp_hal_stdout_tx_strn(const char *str, mp_uint_t len) {
     return did_write ? ret : 0;
 }
 
+void mp_hal_delay_us(mp_uint_t us) {
+    // Avoid calling sleep_us() and invoking the alarm pool by splitting long
+    // sleeps into an optional longer sleep and a shorter busy-wait
+    uint64_t end = time_us_64() + us;
+    if (us > 1000) {
+        mp_hal_delay_ms(us / 1000);
+    }
+    while (time_us_64() < end) {
+        // Tight loop busy-wait for accurate timing
+    }
+}
+
 void mp_hal_delay_ms(mp_uint_t ms) {
-    absolute_time_t t = make_timeout_time_ms(ms);
+    mp_uint_t start = mp_hal_ticks_ms();
+    mp_uint_t elapsed = 0;
     do {
-        mp_event_handle_nowait();
-    } while (!best_effort_wfe_or_timeout(t));
+        mp_event_wait_ms(ms - elapsed);
+        elapsed = mp_hal_ticks_ms() - start;
+    } while (elapsed < ms);
 }
 
 void mp_hal_time_ns_set_from_rtc(void) {
     // Delay at least one RTC clock cycle so it's registers have updated with the most
     // recent time settings.
-    sleep_us(23);
+    mp_hal_delay_us(23);
```

Comment:
Looks like this needs to be 24 us to get the lfs mtime test to pass.

Reply:
`mp_hal_delay_us(24)` doesn't work either, actually. `mp_hal_delay_us(100)` works here, `mp_hal_delay_us(48)` does not.

Measure 76us for `sleep_us(23)` and 25us for `mp_hal_delay_us(23)`. 

Do you remember where 23 comes from? 44kHz?

Reply:
`mp_hal_delay_us(76)` works, too.

Reply:
Ah yeah, 46875Hz. Weird.

I can make this `mp_hal_delay_us(100)` to play it safe and merge this PR, or I can try to figure out why it's more than one nominal RTC clock cycle here?

Reply:
I think it should be 1/32768.

The issue seems to be that your new mp_hal_delay_us is more accurate than the original sleep_us (which is good!).

I originally tuned this 23 value so it was as short as possible while still working.

Reply:
According to this it's 2 RTC clock cycles (1e6/46875*2=43us). I'll re-test 44us to be sure it didn't work.

![image](https://github.com/micropython/micropython/assets/205573/e70ee892-3896-4fa1-a239-404880c083e0)

Reply:
`mp_hal_delay_us(44)` is passing now, maybe 48 didn't fail like I thought (I tested a few different values in a row.)

Reply:
A value of 35 works for me.  I also tested 36 and 40, which both work.  34 does not.

Here's my test script:
```py
def f():
    import time
    ticks_us = time.ticks_us
    sleep_us = time.sleep_us
    # sync to a second
    t = time.time()
    while time.time() == t:
        pass
    # wait to get close to the next second
    sleep_us(950_000)
    # see the next second clock over
    for _ in range(50):
        t0 = ticks_us()
        sleep_us(900)
        t1 = ticks_us()
        print(time.ticks_us(), time.time(), time.time_ns(), t1-t0)
f()

```

Reply:
44 also works for me.  Let's go with that and update the comment to say "delay at least two RTC clock cycles".

Reply:
Done, PTAL.

A cross-over at 35 (actually 37us) is consistent with the other values I tested, too.

It's interesting, my guess is that the other instructions, cache misses, etc. add enough total interval that the RTC clock domain operations order correctly (maybe the write doesn't have to actually complete, but it does need to be ordered before the read). 44 seems like the safest bet, though (particular with unknown caching delays in the experimental tests).